### PR TITLE
implement vectorized builtinCastRealAsDecimalSig

### PR DESCRIPTION
**Pull request by ZhaoQian888**
PCP #12058 
### What problem does this PR solve? 

implement vectorized builtinCastRealAsDecimalSig
### What is changed and how it works?
it improve speed

``` 
BenchmarkVectorizedBuiltinCastFunc/builtinCastRealAsDecimalSig-VecBuiltinFunc-8                     4504            265881 ns/op           64848 B/op       2048 allocs/op

BenchmarkVectorizedBuiltinCastFunc/builtinCastRealAsDecimalSig-NonVecBuiltinFunc-8                  3883            288931 ns/op           92144 B/op       2484 allocs/op
```
### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: expression/builtin_cast_vec.go [hunk 1]
```diff
@@ -756,11 +756,45 @@ func (b *builtinCastStringAsJSONSig) vecEvalJSON(input *chunk.Chunk, result *chu
...
 
 func (b *builtinCastRealAsDecimalSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETReal, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err = b.args[0].VecEvalDecimal(b.ctx, input, buf); err != nil {
```
Comment (wshwsh12) [inline expression/builtin_cast_vec.go]:
```suggestion
	if err = b.args[0].VecEvalReal(b.ctx, input, buf); err != nil {
```

File: expression/builtin_cast_vec.go [hunk 2]
```diff
@@ -756,11 +756,45 @@ func (b *builtinCastStringAsJSONSig) vecEvalJSON(input *chunk.Chunk, result *chu
...
+		}
+	} else {
+		for i := 0; i < n; i++ {
+			if bufreal[i] >= 0 {
+				if err = resdecimal[i].FromFloat64(bufreal[i]); err != nil {
+					return err
+				}
+
+			} else {
+				_, err = types.ProduceDecWithSpecifiedTp(&resdecimal[i], b.tp, b.ctx.GetSessionVars().StmtCtx)
```
Comment (wshwsh12) [inline expression/builtin_cast_vec.go]: In non-vec function, this line will run in all time whatever b.inUnion and val is (of course, without error). But in vec function (your code),  it only runs when b.inUnion is true and bufreal[i] < 0.

Comment (ZhaoQian888) [reply expression/builtin_cast_vec.go]: ok ,I will fix these two problem as soon as possible

File: expression/builtin_cast_vec.go [hunk 3]
```diff
@@ -756,11 +756,36 @@ func (b *builtinCastStringAsJSONSig) vecEvalJSON(input *chunk.Chunk, result *chu
...
+	for i := 0; i < n; i++ {
+		if !b.inUnion || bufreal[i] >= 0 {
+			if err = resdecimal[i].FromFloat64(bufreal[i]); err != nil {
+				return err
+			}
+		} else {
+			_, err = types.ProduceDecWithSpecifiedTp(&resdecimal[i], b.tp, b.ctx.GetSessionVars().StmtCtx)
+			if err != nil {
+				return err
+			}
```
Comment (wshwsh12) [inline expression/builtin_cast_vec.go]: In non-vec function, this line will run in all time whatever b.inUnion and val is (of course, without error). But in vec function (your code), it only runs when the check is false.

Comment (ZhaoQian888) [reply expression/builtin_cast_vec.go]: I see. Is that correct?
```
for i := 0; i < n; i++ {
		if !b.inUnion || bufreal[i] >= 0 {
			if err = resdecimal[i].FromFloat64(bufreal[i]); err != nil {
				return err
			}
		}
		_, err = types.ProduceDecWithSpecifiedTp(&resdecimal[i], b.tp, b.ctx.GetSessionVars().StmtCtx)
		if err != nil {
			return err
		}

	}
```

Comment (wshwsh12) [reply expression/builtin_cast_vec.go]: resdecimal[i], err = *types.ProduceDecWithSpecifiedTp(&resdecimal[i], b.tp, b.ctx.GetSessionVars().StmtCtx)

Comment (wshwsh12) [reply expression/builtin_cast_vec.go]: I think it is ok.

**Pull request merged**